### ISSUE TITLE
spike: Miner Selection Criteria for Validators

### ISSUE THREAD

**Issue by Luka-Loncar** [masa-finance/masa-bittensor#63]
Currently the approach we use to select miners takes a random set of UIDs and does not include a #64 that is required to ensure that every miner doesn't respond to every request on the network. 

## Outcomes

- [ ] #92

## Acceptance criteria

- [ ] We know a Miner selection criteria to adopt
- [ ] We have a follow-up card tracking the implementation

**Comment by mudler** [comment]
This is currently blocked by #64

**Comment by hide-on-bush-x** [comment]
Hey @mudler is there any reason this is blocked by #64 ?
I have the feeling like this two things can go in parallel

**Comment by mudler** [comment]
> Hey @mudler is there any reason this is blocked by #64 ? I have the feeling like this two things can go in parallel

we shouldn't let pick up miners which are blacklisted, hence why we need first to understand how we can handle blacklisting of miners before diving into this one which might depend on logic that we have yet still to discover

**Comment by mudler** [comment]
> Oh, no no, there is a little difference Blacklist: Miners blocking validators requests Selection criteria: Validators selecting what miners to call
> 
> Thats why I believe this can be done in parallel

Sorry, could you expand? it's not entirely clear to me what do you mean here - from the description of the card, we want to clearly avoid picking up miners given a blacklist:

> Currently the approach we use to select miners takes a random set of UIDs and does not include a https://github.com/masa-finance/masa-bittensor/issues/64 that is required to ensure that every miner doesn't respond to every request on the network.

**Comment by hide-on-bush-x** [comment]
Yeah, description seems to be a little bit off

Miners have a blacklist logic to define which validator ignore based on a vpermit ( by default ): https://github.com/masa-finance/masa-bittensor/blob/main/neurons/miner.py#L87

Validators on the other side have this selection criteria, which is used to pick the best miners given some criteria, this ticket is to explore posibilities around that, say "only calling miners that have more than xyz score on chain" "Call miners based on a whitelist" 
That logic will go somewhere in the forwarder close to the UID definition here: https://github.com/masa-finance/masa-bittensor/blob/main/masa/validator/forwarder.py#L31

Both things are similar but not dependant and whatever we find out on the blacklist spike does not impact this task

**Comment by mudler** [comment]
got it, thanks @hide-on-bush-x!

right, this doesn't look really a blocker then, and we can always go back at this if we want to include some sorta of logic around blacklisting miners when validators are going to pick these.

**Comment by hide-on-bush-x** [comment]
Thanks @mudler !

**Comment by juanmanso** [comment]
Moved to Ready and taking the lead for these 2 hours and doing hand-off to @hide-on-bush-x near daily meeting

**Comment by juanmanso** [comment]
# WIP

Since miners are described by their metadata, the first thing we need to inspect is the **Metagraph** to understand which params are key in the evaluation of Miner's eligibility ([link to API docs](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph)).

## Key components of Metagraph Object for Miners
- `ranks` – Neuron rankings as per the Yuma Consensus algorithm, influencing their incentive distribution and network authority ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.R)).
- `trust` – Scores indicating the reliability of neurons, mainly miners, within the network’s operational context ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.T)).
- `consensus` – Scores reflecting each neuron’s alignment with the network’s collective decisions ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.C)).
- `incentive` – Rewards allocated to neurons, particularly miners, for their network contributions ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.I)).
- `dividends` – Rewards received primarily by validators as part of the incentive mechanism ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.D)).
- `active` – Status indicating whether a neuron is actively participating in the network.
- `weights` – Inter-neuronal weights set by each neuron, influencing network dynamics ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.W)).


### Lower importance but could be useful
- `n` (torch.nn.Parameter) – The total number of neurons in the network, reflecting its size and complexity.
- `stake` – Represents the cryptocurrency staked by neurons, impacting their influence and earnings within the network ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.S)).
- `bonds` – Represents speculative investments by neurons in others, part of the reward mechanism ([tensor getter](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.B)).
- `uids` – Unique identifiers for each neuron, essential for network operations.

**Comment by juanmanso** [comment]
## How do other subnets select miners?

- SN 1 / Prompting -> [Random](https://github.com/macrocosm-os/prompting/blob/082c52a4e4667e7044596480fdc59d6c8582f04b/prompting/forward.py#L185)

- SN 8 / Trading Network -> Random: [UUID Tracker](https://github.com/taoshidev/proprietary-trading-network/blob/185dc688b129e496d478c1e84982e2cb869ee552/vali_objects/uuid_tracker.py) + [No Priorization](https://github.com/taoshidev/proprietary-trading-network/blob/185dc688b129e496d478c1e84982e2cb869ee552/neurons/validator.py#L272) ([Usage](https://github.com/taoshidev/proprietary-trading-network/blob/185dc688b129e496d478c1e84982e2cb869ee552/neurons/validator.py#L191))

- SN 9 / Pretraining -> Seems by UID: [1 eval per epoch](https://github.com/macrocosm-os/pretraining/blob/1285929220125e8c9eeaaa8b3a6892c1c8a7ab70/utilities/miner_iterator.py#L9) + [keeps track of pending uids to eval](https://github.com/macrocosm-os/pretraining/blob/1285929220125e8c9eeaaa8b3a6892c1c8a7ab70/neurons/validator.py#L509) + [tracked by filepath ](https://github.com/macrocosm-os/pretraining/blob/1285929220125e8c9eeaaa8b3a6892c1c8a7ab70/neurons/validator.py#L114)

- SN 15 / Blockchain Insights -> [Top miners](https://github.com/blockchain-insights/blockchain-data-subnet/blob/c6b8d4e7e9607d0f5ed2bac24857ab1f11cf56fb/neurons/validators/utils/uids.py#L41) are defined as the ones available (filter ([validators](https://github.com/blockchain-insights/blockchain-data-subnet/blob/c6b8d4e7e9607d0f5ed2bac24857ab1f11cf56fb/neurons/validators/utils/uids.py#L31) + [without IP](https://github.com/blockchain-insights/blockchain-data-subnet/blob/c6b8d4e7e9607d0f5ed2bac24857ab1f11cf56fb/neurons/validators/utils/uids.py#L35) + [not serving](https://github.com/blockchain-insights/blockchain-data-subnet/blob/c6b8d4e7e9607d0f5ed2bac24857ab1f11cf56fb/neurons/validators/utils/uids.py#L24)(aka [not hooked to the network](https://docs.bittensor.com/python-api/html/autoapi/bittensor/axon/index.html#bittensor.axon.axon.serve))) + [ping](https://github.com/blockchain-insights/blockchain-data-subnet/blob/c6b8d4e7e9607d0f5ed2bac24857ab1f11cf56fb/neurons/validators/utils/uids.py#L68C39-L68C48) to check availability) and sorted by [score eligibility](https://github.com/blockchain-insights/blockchain-data-subnet/blob/c6b8d4e7e9607d0f5ed2bac24857ab1f11cf56fb/neurons/validators/utils/uids.py#L78) factoring both [trust](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.T) and [incentive](https://docs.bittensor.com/python-api/html/autoapi/bittensor/metagraph/index.html#bittensor.metagraph.metagraph.I) values.

- SN 16 /  3D Gen -> [Tasks are appended](https://github.com/404-Repo/three-gen-subnet/blob/89d15392a6fc7c84e156b342035763726f6d78ef/neurons/validator/__init__.py#L165) and taken by Miners when available [querying validators](https://github.com/404-Repo/three-gen-subnet/blob/89d15392a6fc7c84e156b342035763726f6d78ef/neurons/miner/validator_selector.py#L29) filtered by [serving status](https://github.com/404-Repo/three-gen-subnet/blob/89d15392a6fc7c84e156b342035763726f6d78ef/neurons/miner/validator_selector.py#L40), [min stake](https://github.com/404-Repo/three-gen-subnet/blob/89d15392a6fc7c84e156b342035763726f6d78ef/neurons/miner/validator_selector.py#L41) and [cooldown periods](https://github.com/404-Repo/three-gen-subnet/blob/89d15392a6fc7c84e156b342035763726f6d78ef/neurons/miner/validator_selector.py#L42).

- SN 20 / BitAgent -> [Notebook on docs](https://github.com/RogueTensor/bitagent_subnet/blob/c676ab4b33a6d087d65f4c7cf347cb49fc9c4b80/docs/tasks.ipynb#L92) selecting by Incentive but code might show it's by [random selection](https://github.com/RogueTensor/bitagent_subnet/blob/c676ab4b33a6d087d65f4c7cf347cb49fc9c4b80/bitagent/validator/forward.py#L51) or given by [Synapse Object](https://github.com/RogueTensor/bitagent_subnet/blob/c676ab4b33a6d087d65f4c7cf347cb49fc9c4b80/bitagent/validator/forward.py#L69).

- SN 22 / Meta Search -> [Random or All ](https://github.com/Datura-ai/smart-scrape/blob/4a8007f5d77d60739778735f9c800966535af2df/datura/__init__.py#L91)on [get_uids](https://github.com/Datura-ai/smart-scrape/blob/4a8007f5d77d60739778735f9c800966535af2df/neurons/validators/validator.py#L165). 

- SN 27 / Compute -> Random but looks like it [filters those axons that share IP.](https://github.com/neuralinternet/compute-subnet/blob/2a2f51ec9f93ac77ada4346cf3a44b08a2dc295c/neurons/validator.py#L374) I infer that they dismiss users that have multiple neurons.

- SN 32 / It's AI -> [Random](https://github.com/It-s-AI/llm-detection/blob/be4417f83e2e2762a28b50b61178962bac8e50af/detection/validator/forward.py#L87)

## Conclusion from this analysis

- Almost every other relevant Subnet (Emissions >~ 1%) is doing random selection
- Those that keep it safe, consider if axons (miners) are available and/or active
- When selecting between top miners, incentive and trust data is taking into account to sort and/or select
- Some others swap selection so that Miners take tasks from validators which are queried based on availability, stake and cooldown periods

**Comment by juanmanso** [comment]
Moving it to `in review` since research looks to be over and its up to us to determine what should be our priority in Miner Selection and take action

**Comment by mudler** [comment]
@juanmanso can you please fill-in https://github.com/masa-finance/masa-bittensor/issues/92 with the implementation details so we can pick it up in the next planning session? thank you!

**Comment by juanmanso** [comment]
> [@juanmanso](https://github.com/juanmanso) can you please fill-in [#92](https://github.com/masa-finance/masa-bittensor/issues/92) with the implementation details so we can pick it up in the next planning session? thank you!

@mudler 
I understood that this ticket needs to define the criteria. However, I just exposed different points of view/implementations but did not choose any. I'd say that depends on the goal of the subnet and the incentive mechanism to choose which selection criteria to take.

Personally, I would choose a random sampling of uids with some minimum filtering to guarantee that selected ones are operational miners but still keep it random. Although if we want to promote activity by checking incentives and/or trust, we could go down that path as well.

Since this is more of a business decision than a technical one from my POV, I left it for review of stakeholders or anyone in position to make the decision. If it's up to me the decision, then I'll complete #92 myself and close this ticket
